Add a good practices page for mutating webhook design (#49626)
kubernetes/website: content/en/docs/concepts/security/mutating-webhooks-good-practices.md

Reviewer: I might put this in the Cluster Administration section, and have https://kubernetes.io/docs/concepts/security/ also link there.

Author: I'll do this before submission so that it doesn't muddle up the reviews

Reviewer: Please do that @shannonxtreme 

[LGTM](https://github.com/kubernetes/website/pull/49626#issuecomment-2741602033) noted but:
/lgtm cancel

Reviewer 2: sorry, I didn't notice this comment

Reviewer: It's fine, we can use that recorded LGTM to skip detailed reviews later.

---

> title: Mutating Webhook Good Practices

Reviewer: Here's the dream:
```suggestion
title: Admission Webhook Good Practices
```

What's the smallest piece of work we can do to make this applicable to both validating and mutating admission webhooks?

Author: I was really leaning towards this as I wrote it. A lot of this is applicable to both, and we can still explicitly call out when advice applies to just mutating. I'll rework wording on the page soon to try and make this agnostic

description: >
  Recommendations for writing mutating admission webhooks in Kubernetes.
content_type: concept
weight: 60
---

<!-- overview -->

This page provides good practices and considerations when designing
_mutating admission webhooks_ in Kubernetes. This information is intended for
cluster operators who run your own admission webhook servers or third-party
applications that modify your API requests.

Before reading this page, ensure that you're familiar with the following
concepts:

* [Admission controllers](/docs/reference/access-authn-authz/admission-controllers/)
* [Admission webhooks](/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks)

<!-- body -->

## Importance of good webhook design {#why-good-webhook-design-matters}

Mutating admission control occurs when any create, update, or delete request
is sent to the Kubernetes API. These webhooks are often written to ensure that
specific fields in object specifications exist or have specific allowed values.

> With every release, Kubernetes adds or modifies the API with new features,

Reviewer 2: this section needs to better match the content of the article and also highlight other risks of webhooks

feature promotions to beta or stable status, and deprecations. Even stable
Kubernetes APIs are likely might change. For example, the `Pod` API changed in
v1.29 to add the
[Sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/) feature.

> Webhooks that worked as expected with earlier versions of an API might not be

Reviewer 2: ```suggestion
It is rare or even not possible that the Kubernetes object that worked before will be broken by introduction of a new API, webhooks that worked as expected with earlier versions of an API might not be
```

Reviewer 2: Trying to highlight the fact that in general upgrades are safe. The issue is that webhooks may be authored badly

able to reconcile more recent changes to that API. This can result in unexpected
behavior after you upgrade your clusters to newer versions.

> ## Identify whether you use mutating webhooks {#identify-mutating-webhooks}

Reviewer: ```suggestion
## Identify whether you use admission webhooks {#identify-admission-webhooks}
```

> 

Reviewer: I suggest: talk about validating webhooks here too.

Even if you don't run your own mutating admission webhooks, some third-party
applications that you run in your clusters might include mutating webhooks.

To check whether your cluster has any mutating webhooks, run the following
command:

```shell
kubectl get mutatingwebhookconfigurations
```

> The output lists any mutating admission controllers in the cluster. 

Reviewer: We used to support configuring these statically by passing a plugin configuration file to the kube-apiserver.

At technical review time **or earlier**, let's validate the the old method is completely omitted. Otherwise, this text isn't 100% accurate.

Author: Is this the bit that's listed here: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers? Or is that different?

## Choose an admission control mechanism {#choose-admission-mechanism}

Kubernetes includes multiple admission control and policy enforcement options.
Knowing when to use a specific option can help you to improve latency and
performance, reduce management overhead, and avoid issues during version
upgrades. The following table describes the mechanisms that let you mutate or
validate resources during admission:

> <table>

Reviewer: Can we use a Markdown table here? If not, add an HTML comment about why it needs to be HTML.

Author: The bulleted lists for use cases make this more readable IMO, but I can also use HTML unordered lists in a markdown table if you prefer that

Reviewer: That reason's fine; what we need now is a brief comment (in the page source) about why, so we don't lose the context.

        policy declarations.</td>
      <td><ul>
        <li>Validate critical configurations before resource admission.</li>
        <li>Enforce complex policy logic before admission.</li>
      </ul></td>
    </tr>
    <tr>
      <td><a href="/docs/reference/access-authn-authz/validating-admission-policy/">Validating admission policy</a></td>
      <td>Intercept API requests before admission and validate against CEL
        expressions.</td>
      <td><ul>
        <li>Validate critical configurations before resource admission.</li>
        <li>Enforce policy logic using CEL expressions.</li>
      </ul></td>
    </tr>
  </tbody>
</table>

In general, use _webhook_ admission control when you want an extensible way to
declare or configure the logic. Use built-in CEL-based admission control when
you want to declare simpler logic without the overhead of running a webhook
server. The Kubernetes project recommends that you use CEL-based admission
control when possible.

### Use built-in validation and defaulting for CustomResourceDefinitions {#no-crd-validation-defaulting}

If you use
{{< glossary_tooltip text="CustomResourceDefinitions" term_id="customresourcedefinition" >}},

> don't use mutating webhooks to validate values in CustomResource specifications
> or to set default values for fields. Kubernetes lets you define validation rules

Reviewer: ```suggestion
don't use admission webhooks to validate values in CustomResource specifications
or to set default values for fields. Kubernetes lets you define validation rules
```

and default field values when you create CustomResourceDefinitions.

To learn more, see the following resources:

* [Validation rules](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)

> * [Defaulting](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting)

Reviewer: If we can link to a concept page here, even better (there may not be a good concept page yet).

Author: We don't. It looks like the "Advanced topics" main section in the task page is mostly conceptual content, so probably a good candidate for moving to a concept page. We have https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/ but that only lightly touches on CRD design

## Performance and latency {#performance-latency}

This section describes recommendations for improving performance and reducing
latency. In summary, these are as follows:

* Consolidate webhooks and limit the number of API calls per webhook.
* Use audit logs to check for webhooks that repeatedly do the same action.
* Use load balancing for webhook availability.
* Set a small timeout value for each webhook.

> ### Improve latency in mutating webhooks {#improve-latency-mutating-webhooks}

Reviewer: ```suggestion
### Design admission webhooks for low latency {#low-latency-design}
```

Mutating admission webhooks are called in sequence. Depending on the mutating
webhook setup, some webhooks might be called multiple times. Every mutating
webhook call adds latency to the admission process. This is unlike validating
webhooks, which get called in parallel. 

When designing your mutating webhooks, consider your latency requirements and
tolerance. The more mutating webhooks there are in your cluster, the greater the
chance of latency increases. 

> Consider the following to reduce latency:

Reviewer 2: Should we advice how to measure this latency? Is there any articles about it already?

@jpbetz do you know?

Reviewer 2: yeah. Also distributed tracing enabled on webhooks will be handy to see what's going on there

that a different controller removes, your webhook gets called again. This leads
to a loop.

To detect these loops, try the following:

1.  Update your cluster audit policy to log audit events. Use the following
    parameters:
    
      * `level`: `RequestResponse`
      * `verbs`: `["patch"]`
      * `omitStages`: `RequestReceived`

    Set the audit rule to create events for the specific resources that your
    webhook mutates.

1.  Check your audit events for webhooks being reinvoked multiple times with the
    same patch being applied to the same object, or for an object having
    a field updated and reverted multiple times.

### Set a small timeout value {#small-timeout}

Admission webhooks should evaluate as quickly as possible (typically in
milliseconds), since they add to API request latency. Use a small timeout for
webhooks.

For details, see
[Timeouts](/docs/reference/access-authn-authz/extensible-admission-controllers/#timeouts).

### Use a load balancer to ensure webhook availability {#load-balancer-webhook}

Admission webhooks should leverage some form of load-balancing to provide high
availability and performance benefits. If a webhook is running within the

> cluster, you can run multiple webhook backends behind a Service to use the

Reviewer: nit: use a glossary tooltip or hyperlink to Service?

Author: I couldnt find info on the Service page that says that there is built-in load balancing that occurs. An external site https://learnk8s.io/kubernetes-long-lived-connections claimed that there's some sort of routing logic that occurs with kube-proxy, but I don't know how true that is

Service load balancing.

## Request filtering {#request-filtering}

This section provides recommendations for filtering which requests trigger
specific webhooks. In summary, these are as follows:

* Limit the webhook scope to avoid system components and read-only requests.
* Limit webhooks to specific namespaces.
* Use match conditions to perform fine-grained request filtering.
* Match all versions of an object.

### Limit the scope of each webhook {#webhook-limit-scope}

> Mutating webhooks run when an API request matches the webhook configuration.

Reviewer: ```suggestion
Admission webhooks are only called when an API request matches the associated webhook configuration.
```

Limit the scope of each webhook to reduce unnecessary calls to the webhook
server. Consider the following scope limitations:

> * Don't match objects in the `kube-system` namespace. If you run your own Pods

Reviewer: ```suggestion
* Avoid matching objects in the `kube-system` namespace. If you run your own Pods
```

  in the `kube-system` namespace, use an

>   [objectSelector](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-objectselector)

Reviewer: ```suggestion
  [`objectSelector`](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-objectselector)
```

  to avoid mutating a critical workload.

> * Don't match node leases. Intercepting node leases might result in failed node
>   upgrades.

Reviewer: ```suggestion
* Don't mutate node leases (Lease objects within the `kube-node-lease` system namespace). Intercepting
  node leases might result in failed node upgrades. Only apply validation controls where you are confident
  that these won't put your cluster at risk; if unsure, exclude the checks.
```

> * Don't match `TokenReview` or `SubjectAccessReview` requests. These are always

Reviewer: ```suggestion
* Don't mutate TokenReview or SubjectAccessReview objects. These are always
```

  read-only requests. Modifying these requests might break your cluster.
* Limit each webhook to a specific namespace by using a

>   [namespaceSelector](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-namespaceselector).

Reviewer: ```suggestion
  [`namespaceSelector`](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-namespaceselector).
```

### Filter for specific requests by using match conditions {#filter-match-conditions}

Admission controllers support multiple fields that you can use to match requests
that meet specific criteria. For example, you can use a `namespaceSelector` to
filter for requests that target a specific namespace.

For more fine-grained request filtering, use the `matchConditions` field in your
webhook configuration. This field lets you write multiple CEL expressions that
must evaluate to `true` for a request to trigger your admission webhook. Using
`matchConditions` might significantly reduce the number of calls to your webhook
server.

For details, see
[Matching requests: `matchConditions`](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchconditions).

> ### Match all versions of an object {#match-all-versions}

Reviewer: ```suggestion
### Match all versions of an API {#match-all-versions}
```

By default, admission webhooks run on any API versions that affect a specified
resource. The `matchPolicy` field in the webhook configuration controls this
behavior. Specify a value of `Equivalent` in the `matchPolicy` field or omit
the field to allow the webhook to run on any API version. 

For details, see
[Matching requests: `matchPolicy`](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchpolicy).

## Mutation scope and field considerations {#mutation-scope-considerations}

This section provides recommendations for the scope of mutations and any special
considerations for object fields. In summary, these are as follows:

* Patch only the fields that you need to patch.
* Don't overwrite array values.
* Avoid side effects in mutations when possible.
* Avoid self-mutations.
* Fail open and validate the final state.
* Plan for future field updates in later versions.
* Prevent webhooks from self-triggering.
* Don't change immutable objects.

### Patch only required fields {#patch-required-fields}

Admission webhook servers send HTTP responses to indicate what to do with a

> specific Kubernetes API request. This response is an `AdmissionReview` object.

Reviewer: AIUI: no backticks for AdmissionReview.

A mutating webhook can add specific fields to mutate before allowing admission
by using the `patchType` field and the `patch` field in the response. Ensure
that you only modify the fields that require a change. 

For example, consider a mutating webhook that's configured to ensure that
`web-server` Deployments have at least three replicas. When a request to

> create a `Deployment` object matches your webhook configuration, the webhook

Reviewer: ```suggestion
create a Deployment matches your admission webhook configuration, the webhook
```

should only update the value in the `spec.replicas` field.

### Don't overwrite array values {#dont-overwrite-arrays}

Fields in Kubernetes object specifications might include arrays. Some arrays
contain key:value pairs (like the `envVar` field in a container specification),
while other arrays are unkeyed (like the `readinessGates` field in a Pod
specification). The order of values in an array field might matter in some
situations. For example, the order of arguments in the `args` field of a
container specification might affect the container. 

Consider the following when modifying arrays:

* Whenever possible, use the `add` JSONPatch operation instead of `replace` to
  avoid accidentally replacing a required value.
* Treat arrays that don't use key:value pairs as sets.
* Ensure that the values in the field that you modify aren't required to be
  in a specific order. 
* Don't overwrite existing key:value pairs unless absolutely necessary.
* Use caution when modifying label fields. An accidental modification might
  cause label selectors to break, resulting in unintended behavior.

### Avoid side effects {#avoid-side-effects}

> Ensure that your webhooks operate only on the content of the `AdmissionReview`

Reviewer: AIUI: no backticks for AdmissionReview.

that's sent to them, and do not make out-of-band changes. These additional
changes, called _side effects_, might cause conflicts during admission if they
aren't reconciled properly. The `.webhooks[].sideEffects` field should
be set to `None` if a webhook doesn't have any side effect. 

If side effects are required during the admission evaluation, they must be
suppressed when processing an `AdmissionReview` object with `dryRun` set to
`true`, and the `.webhooks[].sideEffects` field should be set to `NoneOnDryRun`.

For details, see
[Side effects](/docs/reference/access-authn-authz/extensible-admission-controllers/#side-effects).

### Avoid self-mutations {#avoid-self-mutation}

A webhook running inside the cluster might cause deadlocks for its own
deployment if it is configured to intercept resources required to start its own
Pods.

> For example, a mutating admission webhook is configured to admit `CREATE` Pod

Reviewer: ```suggestion
For example, a mutating admission webhook is configured to admit Pod **create**
```

requests only if a certain label is set in the Pod (such as `"env": "prod"`).

> The webhook server runs in a Deployment that doesn't set the `"env"` label.

Reviewer: double quotes not needed here

When a node that runs the webhook server Pods becomes unhealthy, the webhook
Deployment tries to reschedule the Pods to another node. However, the existing
webhook server rejects the requests since the `"env"` label is unset. As a
result, the migration cannot happen.

Exclude the namespace where your webhook is running with a

> [namespaceSelector](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-namespaceselector).

Reviewer: ```suggestion
[`namespaceSelector`](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-namespaceselector).
```

downtime. 

Let your mutating webhooks "fail open" by setting the `failurePolicy` field to
Ignore. Use a validating controller to check the state of requests to ensure
that they comply with your policies. 

This approach has the following benefits:

* Mutating webhook downtime doesn't affect compliant resources from deploying.
* Policy enforcement occurs during validating admission control.
* Mutating webhooks don't interfere with other controllers in the cluster.

### Plan for future updates to fields {#plan-future-field-updates}

In general, design your webhooks under the assumption that Kubernetes APIs might
change in a later version. Don't write a server that takes the stability of an
API for granted. For example, the release of sidecar containers in Kubernetes
added a `restartPolicy` field to the Pod API. 

### Prevent your webhook from triggering itself {#prevent-webhook-self-trigger}

Mutating webhooks that respond to a broad range of API requests might
unintentionally trigger themselves. For example, consider a webhook that
responds to all requests in the cluster. If you configure the webhook to create

> `Event` objects for every mutation, it'll respond to its own `Event` object

Reviewer: no backticks for names of API kinds

creation requests.

To avoid this, consider setting a unique label in any resources that your
webhook creates. Exclude this label from your webhook match conditions.

### Don't change immutable objects {#dont-change-immutable-objects}

Some Kubernetes objects in the API server can't change. For example, when you
deploy a {{< glossary_tooltip text="static Pod" term_id="static-pod" >}}, the
kubelet on the node creates a 
{{< glossary_tooltip text="mirror Pod" term_id="mirror-pod" >}} in the API
server to track the static Pod. However, changes to the mirror Pod don't
propagate to the static Pod. 

> Don't attempt to mutate these objects during admission. 

Reviewer: - I would mention that mirror pods have the `kubernetes.io/config.mirror` annotation.
- To avoid security problems, consider _also_ enforcing an allow list of namespaces that can contain mirror pods.

I don't know a better way to detect mirror Pods (eg: from their `.status` field)

## Mutating webhook ordering and idempotence {#ordering-idempotence}

This section provides recommendations for webhook order and designing idempotent
webhooks. In summary, these are as follows:

* Don't rely on a specific order of execution.
* Validate mutations before admission.
* Check for mutations being overwritten by other controllers.
* Ensure that the set of mutating webhooks is idempotent, not just the
  individual webhooks.

> ### Don't rely on mutating webhook order {#dont-rely-webhook-order}

Reviewer: ?
```suggestion
### Don't rely on mutating webhook invocation order {#dont-rely-webhook-order}
```

Mutating admission webhooks don't run in a consistent order. Various factors
might change when a specific webhook is called. Don't rely on your webhook
running at a specific point in the admission process. Other webhooks could still
mutate your modified object.

The following recommendations might help to minimize the risk of unintended
changes:

* [Validate mutations before admission](#validate-mutations)
* Use a reinvocation policy to observe changes to an object by other plugins
  and re-run the webhook as needed. For details, see
  [Reinvocation policy](/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy).

> 

Reviewer: Suggestion: to help people build the right mental model, find an excuse to add two validation-only examples.

One should be a delete. The idea in my mind is an immutable ConfigMap with an immutable-until annotation. Until the timestamp in the annotation expires, only exempt callers can modify or delete the ConfigMap.

Why:
1. To illustrate that deletes are subject to admission control
2. To illustrate exemption

The other validation example can be trivial.

   resource limits.

1. For a `CREATE` Pod request, inject a sidecar container with name
   `foo-sidecar` if no container with the name `foo-sidecar` already exists.

In these cases, the webhook can be safely reinvoked, or admit an object that
already has the fields set.

The following examples show non-idempotent mutation logic:

1. For a `CREATE` Pod request, inject a sidecar container with name
   `foo-sidecar` suffixed with the current timestamp (such as
   `foo-sidecar-19700101-000000`).

   Reinvoking the webhook can result in the same sidecar being injected multiple
   times to a Pod, each time with a different container name. Similarly, the
   webhook can inject duplicated containers if the sidecar already exists in
   a user-provided pod.

1. For a `CREATE`/`UPDATE` Pod request, reject if the Pod has label `"env"` set,
   otherwise add an `"env": "prod"` label to the Pod.

   Reinvoking the webhook will result in the webhook failing on its own output.

1. For a `CREATE` Pod request, append a sidecar container named `foo-sidecar`
   without checking whether a `foo-sidecar` container exists.

   Reinvoking the webhook will result in duplicated containers in the Pod, which

>    makes the request invalid and rejected by the API server.

Reviewer: (optional because this content is moved)
The **create** verb is written in bold lowercase, as with all logical API verbs.

sidecar containers. You could run a validating webhook to ensure that those
init containers retained the `restartPolicy: Always` configuration after all
mutations were completed. 

For details, see the following resources:

* [Validating Admission Policy](/docs/reference/access-authn-authz/validating-admission-policy/)
* [ValidatingAdmissionWebhooks](/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook)

## Mutating webhook deployment {#mutating-webhook-deployment}

This section provides recommendations for deploying your mutating admission
webhooks. In summary, these are as follows:

* Gradually roll out the webhook configuration and monitor for issues by
  namespace.
* Limit access to edit the webhook configuration resources. 
* Limit access to the namespace that runs the webhook server, if the server is
  in-cluster.
* Consider cluster availability needs during webhook design.

### Install and enable a mutating webhook {#install-enable-mutating-webhook}

When you're ready to deploy your mutating webhook to a cluster, use the
following order of operations: 

1.  Install the webhook server and start it.

> 1.  Set the `failurePolicy` field in the `MutatingWebhookConfiguration` object
>     to Ignore. This lets you avoid disruptions caused by misconfigured webhooks.
> 1.  Set the `namespaceSelector` field in the `MutatingWebhookConfiguration`
>     object to a test namespace.
> 1.  Deploy the `MutatingWebhookConfiguration` object to your cluster.

Reviewer: ```suggestion
1.  Set the `failurePolicy` field in the MutatingWebhookConfiguration manifest
    to Ignore. This lets you avoid disruptions caused by misconfigured webhooks.
1.  Set the `namespaceSelector` field in the MutatingWebhookConfiguration
    manifest to a test namespace.
1.  Deploy the MutatingWebhookConfiguration to your cluster.
```

Monitor the webhook in the test namespace to check for any issues, then roll the
webhook out to other namespaces. If the webhook intercepts an API request that
it wasn't meant to intercept, pause the rollout and adjust the scope of the
webhook configuration.

### Limit edit access to mutating webhooks {#limit-edit-access}

Mutating webhooks are powerful Kubernetes controllers. Use RBAC or another
authorization mechanism to limit access to your webhook configurations and

> servers. Ensure that the following access is only available to trusted

Reviewer: ```suggestion
servers. For Kubernetes RBAC, ensure that the following access is only available to trusted
```

entities:

> * Verbs: `create`, `update`, `patch`, `delete`, `deletecollection`
> * API group: `admissionregistration.k8s.io/v1`
> * Resources: `MutatingWebhookConfigurations`

Reviewer: ```suggestion
* Verbs: **create**, **update**, **patch**, **delete**, **deletecollection**
* API group: `admissionregistration.k8s.io/v1`
* API kind: MutatingWebhookConfigurations
* ```

If your mutating webhook server runs in the cluster, limit access to create or
modify any resources in that namespace.

### Use a high-availability deployment model {#ha-deployment}

Consider your cluster's availability requirements when designing your webhook. 
For example, during node downtime or zonal outages, Kubernetes marks Pods as
`NotReady` to allow load balancers to reroute traffic to available zones and
nodes. These updates to Pods might trigger your mutating webhooks. Depending on
the number of affected Pods, the mutating webhook server has a risk of timing
out or causing delays in Pod processing. As a result, traffic won't get
rerouted as quickly as you need.

Consider situations like the preceding example when writing your webhooks.
Exclude operations that are a result of Kubernetes responding to unavoidable

> incidents.

Reviewer: Consider moving this near to the load balancer suggestion

## Examples of good implementations {#example-good-implementations}

{{% thirdparty-content %}}

The following projects are examples of "good" custom webhook server
implementations:

* [`cert-manager`](https://github.com/cert-manager/cert-manager/tree/master/internal/webhook)

> * [Gatekeeper Open Policy Agent (OPA)](https://open-policy-agent.github.io/gatekeeper/website/docs/mutation)

Reviewer: These are examples of part of the story. You can take these well-implemented examples and deploy them badly. I wonder if we can convey the nuance here.

## {{% heading "whatsnext" %}}

> * To learn more about RBAC, see the [RBAC documentation](/docs/reference/access-authn-authz/rbac/).

Reviewer: Let's also tell people about webhooks for authz: https://k8s.io/docs/reference/access-authn-authz/webhook/

Maybe also mention declarative admission control (ValidatingAdmissionPolicy & MutatingAdmissionPolicy).

Author: Ah I havent actually updated this section. The RBAC link was from me copying the RBAC good practices markdown file. I'll leave this comment open as a reminder.
